File: DemoServer/AdditionalLanguages/python/textSearch/highlightQueryResultsMapReduce/highlightQueryResultsMapReduce.py
from typing import Optional
from demo_example import RunParamsBase, Example

#region Usings
from ravendb import AbstractIndexCreationTask
from ravendb.documents.indexes.definitions import (
    FieldStorage,
    FieldIndexing,
    FieldTermVector,
)
from ravendb.documents.queries.highlighting import Highlightings, HighlightingOptions
#endregion

class RunParams(RunParamsBase):
    def __init__(
        self,
        search_term: str = None,
        pre_tag: str = None,
        post_tag: str = None,
        fragment_length: int = None,
        fragment_count: int = None,
    ):
        self.search_term = search_term
        self.pre_tag = pre_tag
        self.post_tag = post_tag
        self.fragment_length = fragment_length
        self.fragment_count = fragment_count

#region Demo
#region Step_1
class ArtistsAllSongs(AbstractIndexCreationTask):
#endregion
    #region Step_2
    class IndexEntry:
        def __init__(self, Artist: str = None, AllSongTitles: str = None):
            self.Artist = Artist
            self.AllSongTitles = AllSongTitles
    #endregion

    def __init__(self):
        #region Step_3
        super().__init__()
        self.map = (
            "docs.LastFms.Select(song => new { "
            "    Artist = song.Artist, "
            "    AllSongTitles = song.Title "
            "})"
        )        
        #endregion
        
        #region Step_4
        self.reduce = (
            "results.GroupBy(result => result.Artist).Select(g => new {"
            "    Artist = g.Key,"
            "    AllSongTitles = String.Join(' ', g.Select(x => x.AllSongTitles))"
            "})"
        )
        #endregion

        #region Step_5
        self._store("Artist", FieldStorage.YES)
        
        self._store("AllSongTitles", FieldStorage.YES)
        self._index("AllSongTitles", FieldIndexing.SEARCH)
        self._term_vector("AllSongTitles", FieldTermVector.WITH_POSITIONS_AND_OFFSETS)
        #endregion
#endregion

class HighlightQueryResultsMapReduce(Example):
    def run(self, run_params: RunParams):
        searchTerm = run_params.search_term or "smile"
        preTag = run_params.pre_tag or " (: "
        postTag = run_params.post_tag or " :) "
        fragmentLength = run_params.fragment_length or 80
        fragmentCount = run_params.fragment_count or 1
        ArtistsAllSongs().execute(self.document_store_holder.store())
        
        #region Demo
        highlightings_info: Optional[Highlightings] = None

        def __highlightings_callback(highlight: Highlightings):
            nonlocal highlightings
            highlightings = highlight

        with self.document_store_holder.store().open_session() as session:
            #region Step_6
            highlighting_options = HighlightingOptions()
            highlighting_options.group_key = "Artist"
            highlighting_options.pre_tags = [preTag]
            highlighting_options.post_tags = [postTag]
            #endregion

            #region Step_7
            artists_results = list(
                session.query_index_type(ArtistsAllSongs, ArtistsAllSongs.IndexEntry)
                .highlight(
                    "AllSongTitles",
                    fragmentLength,
                    fragmentCount,
                    __highlightings_callback,
                    highlighting_options,
                )
                .search("AllSongTitles", searchTerm)
            )
            #endregion

            #region Step_8
            if len(artists_results) > 0:
                songs_fragments = highlightings_info.get_fragments(artists_results[0].Artist)
            #endregion
            #endregion

            highlight_results = []
            for artist_item in artists_results:
                songs_fragments = highlightings_info.get_fragments(artist_item.Artist)
                for fragment in songs_fragments:
                    item_results = DataToShow(artist_item.Artist, fragment)
                    highlight_results.append(item_results)

        return sorted(highlight_results, key=lambda x: x.Artist)

class DataToShow:
    def __init__(self, artist: str = None, song_fragment: str = None):
        self.artist = artist
        self.song_fragment = song_fragment

File: DemoServer/AdditionalLanguages/php/textSearch/highlightQueryResultsMapReduce/HighlightQueryResultsMapReduce.php
<?php

namespace RavenDB\Demo\textSearch\highlightQueryResultsMapReduce;

//region Usings
use RavenDB\Documents\Queries\Highlighting\HighlightingOptions;
use RavenDB\Documents\Queries\Highlighting\Highlightings;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\FieldIndexing;
use RavenDB\Documents\Indexes\FieldStorage;
use RavenDB\Documents\Indexes\FieldTermVector;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;

//region Demo
//region Step_1
class ArtistsAllSongs extends AbstractIndexCreationTask
//endregion
{
    public function __construct()
    {
        parent::__construct();

        //region Step_2
        $this->map = "docs.LastFms.Select(song => new {" .
            "    Artist = song.Artist," .
            "    AllSongTitles = song.Title" .
            "})";
        //endregion

        //region Step_3
        $this->reduce = "results.GroupBy(result => result.Artist).Select(g => new {" .
            "    Artist = g.Key," .
            "    AllSongTitles = String.Join(\" \", g.Select(x => x.AllSongTitles))" .
            "})";
        //endregion

        //region Step_4
        $this->store("Artist", FieldStorage::yes());

        $this->store("AllSongTitles", FieldStorage::yes());
        $this->index("AllSongTitles", FieldIndexing::search());
        $this->termVector("AllSongTitles", FieldTermVector::withPositionsAndOffsets());
        //endregion
    }
}
//endregion

//region Demo
//region Step_5
class IndexEntry
{
    private ?string $artist = null;
    private ?string $allSongTitles = null;

    public function getArtist(): ?string
    {
        return $this->artist;
    }

    public function setArtist(?string $artist): void
    {
        $this->artist = $artist;
    }

    public function getAllSongTitles(): ?string
    {
        return $this->allSongTitles;
    }

    public function setAllSongTitles(?string $allSongTitles): void
    {
        $this->allSongTitles = $allSongTitles;
    }
}
//endregion
//endregion

class HighlightQueryResultsMapReduce
{
    public function __invoke(RunParams $runParams): array
    {
        $searchTerm = $runParams->getSearchTerm() ?? "smile";
        $preTag = $runParams->getPreTag() ?? " (: ";
        $postTag = $runParams->getPostTag() ?? " :) ";
        $fragmentLength = $runParams->getFragmentLength() ?? 80;
        $fragmentCount = $runParams->getFragmentCount() ?? 1;

        $highlightingsInfo = new Highlightings();

        //region Demo
        $artistsResults = [];

        $session = DocumentStoreHolder::getMediaStore()->openSession();
        try {
            //region Step_6
            $highlightingOptions = new HighlightingOptions();
            $highlightingOptions->setGroupKey("Artist");
            $highlightingOptions->setPreTags([ $preTag ]);
            $highlightingOptions->setPostTags([ $postTag ]);
            //endregion

            //region Step_7
            $artistsResults = $session->query(IndexEntry::class, ArtistsAllSongs::class)
                ->highlight("AllSongTitles", $fragmentLength, $fragmentCount, $highlightingOptions, $highlightingsInfo)
                ->search("AllSongTitles", $searchTerm)
                ->toList();
            //endregion

            //region Step_8
            if (count($artistsResults) > 0) {
                $songsFragments = $highlightingsInfo->getFragments($artistsResults[0]->getArtist());
            }
            //endregion
        } finally {
            $session->close();
        }
        //endregion

        $highlightResults = [];

        /** @var IndexEntry $artistItem */
        foreach ($artistsResults as $artistItem) {
            $songsFragments = $highlightingsInfo->getFragments($artistItem->getArtist());
            foreach ($songsFragments as $fragment) {
                $itemResults = new DataToShow();
                $itemResults->setArtist($artistItem->getArtist());
                $itemResults->setSongFragment($fragment);

                $highlightResults[] = $itemResults;
            }
        }

        usort($highlightResults, function(DataToShow $a, DataToShow $b) {
            return $a->getArtist() > $b->getArtist();
        });

        return $highlightResults;
    }
}

class DataToShow
{
    private ?string $artist = null;
    private ?string $songFragment = null;

    public function getArtist(): ?string
    {
        return $this->artist;
    }

    public function setArtist(?string $artist): void
    {
        $this->artist = $artist;
    }

    public function getSongFragment(): ?string
    {
        return $this->songFragment;
    }

    public function setSongFragment(?string $songFragment): void
    {
        $this->songFragment = $songFragment;
    }
}
